processor/src/tests/exec_iters.rs

```diff
@@ -0,0 +1,91 @@
+use super::super::build_test;
+use crate::{Felt, FieldElement, VmState};
+
+// EXEC ITER TESTS
+
+#[test]
+fn test_memory() {
+    let source = "begin popw.mem.1 popw.mem.3 end";
+    let test = build_test!(source, &[1, 2, 3, 4, 5, 6, 7, 8]);
+    let traces = test.execute_iter();
+    let mems = [
+        vec![(1_u64, slice_to_word(&[5, 6, 7, 8]))],
+        vec![
+            (1_u64, slice_to_word(&[5, 6, 7, 8])),
+            (3_u64, slice_to_word(&[1, 2, 3, 4])),
+        ],
+    ];
+
+    let stacks = [
+        slice_to_felts(&[8, 7, 6, 5, 4, 3, 2, 1]),
+        slice_to_felts(&[4, 3, 2, 1]),
+    ];
+    // Verify that memory is being written into addresses 1 and then 2 with the right values.
+    for (clk, t) in traces.enumerate() {
+        let state = t.as_ref().unwrap();
+        assert_eq!(clk, state.clk);
+        match state.memory.len() {
+            0 => {}
+            1 => assert_eq!(mems[0], state.memory),
+            2 => assert_eq!(mems[1], state.memory),
+            _ => panic!("Too many memory writes"),
+        }
+        let stack = get_stack_values(state);
+        match stack.len() {
+            8 => assert_eq!(stacks[0], stack),
+            4 => assert_eq!(stacks[1], stack),
+            _ => {}
+        }
+    }
```

Comment:
Could we change this to iterate through trace states explicitly? I.e. using `.next()` rather than a loop. Then, we could check the state at each next clock cycle.

Reply:
fixed